cmd/codegen/generator/go/templates/module_objects.go

```diff
@@ -0,0 +1,226 @@
+
+		fieldSpec.name = name
+
+		// override the name with the json tag if it was set - otherwise, we
+		// end up asking for a name that we won't unmarshal correctly
+		tag := reflect.StructTag(t.Tag(i))
+		if dt := tag.Get("json"); dt != "" {
+			dt, _, _ = strings.Cut(dt, ",")
+			if dt == "-" {
+				continue
+			}
+			fieldSpec.name = dt
+		}
+
+		spec.fields = append(spec.fields, fieldSpec)
+	}
+
+	if ps.isMainModuleObject(spec.name) && ps.constructor != nil {
+		spec.constructor, err = ps.parseGoFunc("", ps.constructor)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse constructor: %w", err)
+		}
+	}
+
+	return spec, nil
+}
+
+type parsedObjectType struct {
+	name string
+	doc  string
+
+	fields      []*fieldSpec
+	methods     []*funcTypeSpec
+	constructor *funcTypeSpec
+}
+
+func (spec *parsedObjectType) TypeDefCode() (*Statement, error) {
+	withObjectArgsCode := []Code{
+		Lit(spec.name),
+	}
+	withObjectOptsCode := []Code{}
+	if spec.doc != "" {
+		withObjectOptsCode = append(withObjectOptsCode, Id("Description").Op(":").Lit(strings.TrimSpace(spec.doc)))
+	}
+	if len(withObjectOptsCode) > 0 {
+		withObjectArgsCode = append(withObjectArgsCode, Id("TypeDefWithObjectOpts").Values(withObjectOptsCode...))
+	}
+
+	typeDefCode := Qual("dag", "TypeDef").Call().Dot("WithObject").Call(withObjectArgsCode...)
+
+	for _, method := range spec.methods {
+		fnTypeDefCode, err := method.TypeDefCode()
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert method %s to function def: %w", method.name, err)
+		}
+		typeDefCode = dotLine(typeDefCode, "WithFunction").Call(Add(Line(), fnTypeDefCode))
+	}
+
+	for _, field := range spec.fields {
+		fieldTypeDefCode, err := field.typeSpec.TypeDefCode()
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert field type: %w", err)
+		}
+		withFieldArgsCode := []Code{
+			Lit(field.name),
+			fieldTypeDefCode,
+		}
+		if field.doc != "" {
+			withFieldArgsCode = append(withFieldArgsCode,
+				Id("TypeDefWithFieldOpts").Values(
+					Id("Description").Op(":").Lit(field.doc),
+				))
+		}
+		typeDefCode = dotLine(typeDefCode, "WithField").Call(withFieldArgsCode...)
+	}
+
+	if spec.constructor != nil {
+		fnTypeDefCode, err := spec.constructor.TypeDefCode()
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert constructor to function def: %w", err)
+		}
+		typeDefCode = dotLine(typeDefCode, "WithConstructor").Call(Add(Line(), fnTypeDefCode))
+	}
+
+	return typeDefCode, nil
+}
+
+func (spec *parsedObjectType) GoSubTypes() []types.Type {
+	var subTypes []types.Type
+	for _, method := range spec.methods {
+		subTypes = append(subTypes, method.GoSubTypes()...)
+	}
+	for _, field := range spec.fields {
+		subTypes = append(subTypes, field.typeSpec.GoSubTypes()...)
+	}
+	if spec.constructor != nil {
+		subTypes = append(subTypes, spec.constructor.GoSubTypes()...)
+	}
+	return subTypes
+}
+
+type fieldSpec struct {
+	name     string
+	doc      string
+	typeSpec ParsedType
+}
```

Comment:
WDYT about idea of merging some of the parts of this into `paramSpec` to create a base `spec` type of some variety?

The devil is in the details for sure, but conceptually this makes sense to me:
- They're both parsed from go field parts of the AST
- The way their pragmas are extracted is identical (though some of the meaning is a bit different)
  - E.g. instead of being a property, `optional` could be a method that checks the `pragmas` property
  - E.g. instead of being a property, `description` could be a method that selects the correct doc comment or line comment
- At some point, I know we've discussed what `Optional`/`// +optional` might look like on struct fields - if we implement the CLI behavior to allow flattening structs out to be parameters, then this would make that struct optional. This one is a bit more of a stretch though.

Then `parseParamSpecVar` could be used to parse `fieldSpecs` as well.

Reply:
I tried this but I'm leaning towards leaving this as is and saving for further refactorization as we continue modifying this code in the future. Mainly because I think to do this fully+comprehensibly would require another significant chunk of changes and I'm trying to keep this very scoped so I don't get overly distracted for too long from the main goal of interface support (and everything else) :-)

I agree we'd want some common base `spec` thing that is embedded by one type for function args and one type for struct fields; otherwise you get weird things like some of the fields in the spec having different meaning/validity for whether it's an arg vs field and the `graphqlName` method needing to be different for each of them.

Let me know if that sounds good to you!

Reply:
Yeah, all good points :tada:

I think it might make sense to do in the future, but at the moment, agreed, interfaces are the thing to aim for :dart: